### PR TITLE
Suggested audio handling in RV session

### PR DESCRIPTION
**Pull request by apetrynet**
Hi!

Here is a suggestion for a way to handle audio in RV session files. The Idea is to treat audio clips as layers instead of separate audio clips. This is done by gathering the necessary metadata needed to add the audio file in extern_rv.py and storing it in the video clips' media reference.

This is a proof of concept that still needs work to make sure it performs correctly with more complex timelines, but simple video over audio seem to work.

My Motivation for this is that I was working on a playblast tool for Maya that required sound. I first tried constructing an otio timeline with audio and video but RV complains about the audio track not having any video. It worked if I switched stack in RV, but seems like this is not the intended way of doing audio in RV. Some research led me to the layer approach.

Please let me know what you guys think of the idea.

### CODE REVIEW

**Comment by rogergodspeed** [comment]
I took a look at this PR and have some thoughts. I see how this strategy marries audio files with other imagery sources by inserting the relevant fragments of audio into the imagery file sources. That is one approach, but I feel like a more OTIO analogous approach would be to add invisible movieprocs to the audio sources and allow them to exist on their own.

OTIO and RV both support arbitrary layers of stacks and sequences/tracks. I think a solution that keeps the two structures more aligned would be preferable in the long term.

You can see how the movieproc is already used in extern_rv.py's _create_media_reference. Also the transition_test.rv file contains a blank movieproc.

How does that sound to you?

**Comment by apetrynet** [comment]
Hi!
Thanks for the feedback. This sounds like a good idea. I'll give it a go.

**Comment by apetrynet** [comment]
Hi @rogergodspeed 
I refactored my code to match your comment and agree that this is a much better way of doing this. However, I came across a hurdle when attempting to figure out if a Clip resides in an audio track or not. The call to `otio.algorithms.track_with_expanded_transitions` in the `_write_track` function returns a list instead of a Track which breaks the "item.parent" connection. I filed an issue #371 on this which I'm awaiting responses to. 
In the meantime, do you have any other suggestions on how to tell if a Clip is an audio clip or not? (other than file extension) 
And do you have any thoughts on whether or not to ignore audio clips that point to a video clip's source? 
I did some quick testing in RV to see if it blows the sound when having the same audio in both video and audio track. Seems like it gets evened out and doesn't cause a problem.

I'm waiting for response on the above mentioned issue before pushing my refactor because it depends on the change I suggested there.

**Comment by apetrynet** [comment]
Hello again!

No need for apologies, I was struggling to express myself clearly. I'll try again :)

I refactored my code following your comments and now create a blank movieproc to accompany an audio clip.
My solution for determining if I am dealing with an audio clip or a video clip is to call the clips `parent()` method. and check if it is a audio track. The problem is that an earlier call to `otio.algorithms.track_with_expanded_transitions()` in the `extern_rv._write_track` method replaces the passed track object with a list. This breaks the link between Clips and their parent Tracks and I have no way of catching audio clips.
I found a way to fix this which is what I submitted a separate issue for.

So my first question I tried to formulate was if you had any ideas for alternative ways of catching an audio clip from the OTIO json in case my solution for the list vs track issue is shot down and I have to refactor my RV related code again. 

My second question was your thought on whether or not to include audio clips in the RVSession file in cases where the audio clip is the same file as the video clip. (QuickTimes with audio for instance) I was worried that two instances of the sound would raise the volume/intensity in RV out of proportions. This because RV would playback the sound embedded in the QuickTime and yet another clip added by the otio RV adapter. I did some testing and this doesn't seem to be a problem.

So to sum up. I have successfully been able to create an rv session file through the adapter with audio in RVFileSources with blank media.
This however is solved with a few tweaks to the track_algo.py file talked about in the #371 issue.

Please ask if I'm still unclear. Thank you so much for your time.

**Comment by rogergodspeed** [comment]
I see! Thank you for spelling that out for me. I am taking another look at your second draft now.

**Comment by ssteinbach** [comment]
Thanks for this @apetrynet!  Is this still [WIP] or are you ready for reviews?

**Comment by apetrynet** [comment]
Hey!
Thanks for the feedback.
This should work for my case. I created a session file that looks the part, but I don't have RV on my home computer unfortunately so I won't be able to confirm for certain util tomorrow morning (my time).

**Comment by rogergodspeed** [comment]
I was going to make one more suggestion about the blank movieprocs. There is another attribute you can use called `filename` that lets you spoof the name of the movieproc as seen in the session manager. Unfortunately I broke that feature in one of my last updates to RV before leaving the company. Therefore once you prove your changes work for you I feel confident signing off on this addition. Seems really helpful.

Thanks!

File: opentimelineio_contrib/adapters/extern_rv.py [hunk 1]
```diff
@@ -207,8 +211,23 @@ def _write_timeline(tl, to_session):
 def _create_media_reference(mr, to_session):
```
Comment (avrata) [inline opentimelineio_contrib/adapters/extern_rv.py]: Not your change, but this `mr` here threw me for quite a loop as it's not really a media reference itself, but something that might contain one. Nothing to change, just something I'll ask about updating another time.

**Comment by apetrynet** [comment]
Hi all!

I can confirm that the latest push works like charm! 
@avrata Would you like me to change the `mr` to `item` or should this wait for other updates?

File: opentimelineio_contrib/adapters/extern_rv.py [hunk 2]
```diff
@@ -207,8 +211,23 @@ def _write_timeline(tl, to_session):
...
+
+            if mr.parent() and mr.parent().kind == otio.schema.TrackKind.Audio:
+                # Create blank video media to accompany audio for valid source
+                blank = "{},start={},end={},fps={}.movieproc".format(
+                    "blank",
+                    mr.available_range().start_time.value,
+                    mr.available_range().end_time_inclusive().value,
+                    mr.available_range().duration.rate
+                )
+                media = [blank, str(mr.media_reference.target_url)]
```
[the remark this answers is not in the document]
Comment (apetrynet) [reply opentimelineio_contrib/adapters/extern_rv.py]: Sure! I tested switching places on a previously created session file and it work nicely. I'll update now.

[the remark this answers is not in the document]
Comment (rogergodspeed) [reply opentimelineio_contrib/adapters/extern_rv.py]: Great! Did you notice how switching the order preserves the name of the audio source in the Session Manager? It hides the movieproc (unless you drill into the source), but promotes the relevant naming information.

There is supposed to be a way to achieve that in movieprocs to begin with, but one of my last significant additions to RV before my departure created a bug in this movieproc feature. I have started a conversation with Shotgun/AutoDesk to see if I can get them to fix my bug, but reordering here wasn't so bad.

File: opentimelineio_contrib/adapters/extern_rv.py [hunk 3]
```diff
@@ -207,8 +211,21 @@ def _write_timeline(tl, to_session):
 def _create_media_reference(mr, to_session):
     if hasattr(mr, "media_reference") and mr.media_reference:
         if isinstance(mr.media_reference, otio.schema.ExternalReference):
-            to_session.setMedia([str(mr.media_reference.target_url)])
+            media = [str(mr.media_reference.target_url)]
+            
+            if mr.parent() and mr.parent().kind == otio.schema.TrackKind.Audio:
+                # Create blank video media to accompany audio for valid source
```
Comment (ssteinbach) [inline opentimelineio_contrib/adapters/extern_rv.py]: can you also add a note here for the fact that putting it *after* the audio means that it puts the correct name onto the RVFileSource?

**Comment by avrata** [comment]
As for "mr" -> "item" in _create_media_reference() ... yes, go for renming that as it'll help future readers of the code not get stuck like I did. Thanks! 

Super excited about this change as it's something we have been running into as well for a while and glad to see this approach used to help make better sense of audio in RV.

File: opentimelineio_contrib/adapters/extern_rv.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (apetrynet) [reply opentimelineio_contrib/adapters/extern_rv.py]: Yeah, that read much nicer. I thought for some reason that the order was important for the composition to work properly. Thanks for the tip.

File: opentimelineio_contrib/adapters/extern_rv.py (hunk 3 above)
Comment (apetrynet) [reply opentimelineio_contrib/adapters/extern_rv.py]: Sure

**Comment by apetrynet** [comment]
Cool! I'll get on it!

File: opentimelineio_contrib/adapters/extern_rv.py (hunk 3 above)
Comment (apetrynet) [reply opentimelineio_contrib/adapters/extern_rv.py]: Added comment in latest commit

**Comment by ssteinbach** [comment]
@apetrynet this is looking good!  When you think its ready, remove the [WIP] prefix from the name of the PR.  Thanks for your contribution!

**Comment by apetrynet** [comment]
Yes, I believe so.

**Comment by GatesARC** [comment]
@rogergodspeed @apetrynet 
The only concern I would have here is with the movieproc resolution.  Typically when viewing in the context of other objects and placing movieprocs, I need to modify the image resolution to match the surrounding objects or else it crops the sequence viewsize to the smallest image size.  I might also be incorrect, I thought it would just be good to double check.

**Comment by apetrynet** [comment]
Hmmm. I didn't experience this when viewing a playblast with audio the other day. Perhaps the blank movieproc doesn't affect other objects like that? Perhaps @rogergodspeed has some insight? I'm at home and not able to get the GUI of RV to run on my laptop, so I won't be able to check until tomorrow again.

**Comment by rogergodspeed** [comment]
I would need an example to see what you are describing @GatesARC, but there is certainly some history of bad behavior in the pursuit of resolution independence. Not being in front of the code anymore or really using RV very often I would have to say if you see anything bug like happening it should certainly be reported to Shotgun/Autodesk. My understanding, dated as it may be, is this PR approach should be supported.

**Comment by jminor** [comment]
Thanks @apetrynet :)

File: opentimelineio_contrib/adapters/extern_rv.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (avrata) [reply opentimelineio_contrib/adapters/extern_rv.py]: Hi @rogergodspeed and @apetrynet, bringing up this old pull request due to a weird edge case we came across here related to this. In a few instances, folks are setting a movie file in the "Audio" track, however this movie brings along some old reference video content. Depending on the stack setup, this reference video can show through. 

I am trying a few approaches to deal with this, but one simple solution I found was to swap the order here back to `media = [blank, str(mr.media_reference.target_url)]`. This seems "mute" the video (ie show only black) when the media_reference is pointing to a movie. We have had a few conversations on the intent of the "Audio" track kind, but one could argue that it should ONLY be for audio content, and ignoring (via "blanking" out) any video content is appropriate. 

This does bring back the question of the name in the session manager. We are getting around that by setting the clip item's name to something that makes sense for us, but I wonder if this code:

https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/master/opentimelineio_contrib/adapters/extern_rv.py#L270

Would cause the ui.name to always be set (and not letting the "blank" movieproc though). 

In anycase, I put in #426 as a possible idea for switching the order again. Feel free to try and talk me out of this, or if you have alternate ideas, I would love to hear them.

Thanks,
Robyn

[the remark this answers is not in the document]
Comment (apetrynet) [reply opentimelineio_contrib/adapters/extern_rv.py]: Hi @avrata !

I unfortunately don't have enough knowledge of the inside workings of RV, so I can't really come up with other reasons why this would be a problem other than readability. 

That said, I think this edge case points at some other issues with OTIO that perhaps should be discussed. If I'm not mistaken, when looking at the media_reference for an audio item based on a movie file it points to a movie file and not an audio file or an audio track within that movie file. This might indicate that we should see this edge case quite often. One might solve this in a media linker, but when exporting from an NLE you often get the movie source as a media reference and most likely will run into this issue. 
There has been some discussion (#343) regarding a way to group media. Perhaps this is a good case for looking into this again?

[the remark this answers is not in the document]
Comment (jminor) [reply opentimelineio_contrib/adapters/extern_rv.py]: You are correct. Currently only the track kind specifies the type of media to expect. If a clip's media reference points to something of the wrong type, or a movie with both in it, the _intent_ is that only the media that matches the parent track's kind should be relevant. It would be reasonable for an adapter to mute any audio found in a video track, and to omit any video found in an audio track. However, since RV is happy to play both that doesn't really work well in practice. OTIO is missing a way for a media reference to select just one channel/stream within a media file - if we had that (and maybe an track kind of "Video+Audio"), then we could make this distinction clear.

[the remark this answers is not in the document]
Comment (rogergodspeed) [reply opentimelineio_contrib/adapters/extern_rv.py]: Every RVFileSource has a `group.noMovieAudio` property (http://www.tweaksoftware.com/static/documentation/rv/current/html/rv_reference.html#RVFileSource), but nothing analogous for video. The main concern I have with "muting" video this way is that it may not be respected when the compositing mode is changed. Anyway I am increasingly less and less familiar with how compositing is handled now that I have been gone over a year. I am taking steps to loop in someone that is still involved.

[the remark this answers is not in the document]
Comment (avrata) [reply opentimelineio_contrib/adapters/extern_rv.py]: Thanks. I did reach out to Kessler about this and he said they could turn the "onlyMovieAudio" (or whatever the inverse of the noMovieAudio flag is) into a feature request. This "blank'ing" of the video seemed like what would work now, but he did offer a warning to maybe do it twice to handle stereo modes.

[the remark this answers is not in the document]
Comment (mikepkes) [reply opentimelineio_contrib/adapters/extern_rv.py]: Hey all,

@avrata did reach out to me and @rogergodspeed just put me in touch with @jminor who pointed me here; sorry I didn't see this sooner!  With some discussion about the original problem with Jon, it sounds like the crux of the issue comes down to how the filename is being populated.

You should be able to use one (or two for stereo) blank movieprocs with the audio movie as the last item.  When you encode the filename, however, you will want to base64-encode the filename; once you do that it should be able to carry with it any name you wish.  If the filename looks like a path, the ui in the session manager will just show the extension-less basename (just as it would if it were a normal file path).

With base64 encoding, you can include all manner of characters that will pass fluidly through the filename attribute on the movieproc:

>>> import base64;base64.b64encode("Tr0ub4dor&3")
'VHIwdWI0ZG9yJjM='

```
/Applications/RV64.app/Contents/MacOS/RV64 blank,filename=VHIwdWI0ZG9yJjM=.movieproc
```

screen shot 2019-02-20 at 5 49 33 pm


Or with paths...

>>> import base64;base64.b64encode("/correct/horse/battery/staple.mov")
'L2NvcnJlY3QvaG9yc2UvYmF0dGVyeS9zdGFwbGUubW92'

```
/Applications/RV64.app/Contents/MacOS/RV64 blank,filename=L2NvcnJlY3QvaG9yc2UvYmF0dGVyeS9zdGFwbGUubW92.movieproc
```

screen shot 2019-02-20 at 5 53 11 pm


http://www.tweaksoftware.com/static/documentation/rv/current/html/rv_manual.html#File_Formats_Movie_File_Formats_RV_s_movieproc_Format___movieproc_


Between this and the using the blank movieproc first in the sources, I think you should be able to get both an audio-only placeholder along with a deterministic display name.
